src/main.rs

```diff
@@ -0,0 +1,482 @@
+        .with_max_sample_rate();
+
+    let config = supported_config.config();
+    let (sender, reciever) = std::sync::mpsc::channel::<f32>();
+    let stream = device
+        .build_input_stream(
+            &config,
+            move |data: &[f32], _: &cpal::InputCallbackInfo| {
+                // samples from different channels are received interleaved
+                // mix all the channels into single one by averaging
+                for b in data.chunks_exact(config.channels as usize) {
+                    let val = b.iter().sum::<f32>() / config.channels as f32;
+                    sender.send(val).unwrap();
+                }
+            },
+            move |err| panic!("failed to get microphone input: {}", err),
+            None,
+        )
+        .unwrap();
+    stream.play().unwrap();
+
+    let mut samples = Vec::new();
+    let mut min_amount = BLOCK_SIZE.max(FFT_SIZE);
+
+    let mut state = State::default();
+    state.sample_rate = config.sample_rate.0 as f32;
+    let mut avg_f0 = 0.0;
+    let mut avg_db = 0.0;
+
+    let fft = FftPlanner::<f32>::new().plan_fft_forward(FFT_SIZE);
+
+    while state.running {
+        while let Ok(sample) = reciever.try_recv() {
+            assert!(f32::abs(sample) <= 1.0);
+            samples.push(sample);
+        }
+
+        if samples.len() >= min_amount {
+            let mut fft_buf = samples[min_amount - FFT_SIZE..min_amount]
+                .iter()
+                .map(|&x| Complex { re: x, im: 0.0 })
+                .collect::<Vec<_>>();
+
+            if state.apply_window {
+                apply_window(&mut fft_buf);
+            }
+
+            fft.process(&mut fft_buf);
+
+            let cmndf = compute_cmndf(&samples[min_amount - BLOCK_SIZE..min_amount]);
+
+            let f0 = compute_f0(&cmndf, state.sample_rate);
+            avg_f0 = f32::exp(f32::ln(avg_f0) + 0.2 * (f32::ln(f0 / avg_f0)));
+            if !avg_f0.is_finite() {
+                avg_f0 = f0;
+            }
+
+            let rms = f32::sqrt(
+                samples[min_amount - FFT_SIZE..min_amount]
+                    .iter()
+                    .map(|x| x * x)
+                    .sum::<f32>()
+                    / BLOCK_SIZE as f32,
+            );
+
+            let db = 20.0 * f32::log10(rms);
+            avg_db = avg_db + 0.4 * (db - avg_db);
+            if !avg_db.is_finite() {
+                avg_db = db;
+            }
+
+            let (note_index, cents) = match state.target_frequency {
+                Some(target_frequency) => {
+                    let cents = get_cents(avg_f0, target_frequency);
+                    let (target_note, _) = get_note(target_frequency);
+                    (target_note, cents)
+                }
+                None => {
+                    let (note, exact_frequency) = get_note(avg_f0);
+                    let cents = get_cents(avg_f0, exact_frequency);
+                    (note, cents)
+                }
+            };
```

Comment:
```suggestion
                    let (note_index, exact_frequency) = get_note(state.target_frequency.unwrap_or(avg_f0));
                    let cents = get_cents(avg_f0, state.target_frequency.unwrap_or(exact_frequency));
```

I'm not sure, maybe you can use this